### DISCUSSION TITLE
How to use attentionentropyloss

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#183] [Archived]
<USER>>>> Had<br/> <DATE> [April 14, 2020, 2:05pm]<br/> <br/> I have found it here *slash* <br/> <br/> [github.com](https://github.com/mozilla/TTS/blob/master/layers/losses.py#L86)<br/> <br/> #### [mozilla/TTS/blob/master/layers/losses.py *slash* #L86](https://github.com/mozilla/TTS/blob/master/layers/losses.py#L86)<br/> <br/> ```<br/> x curl-run-all.sh discourse.mozilla.org html-to-markdown.sh ordered-posts ordered-posts~ TTS.cdx tts.commands tts-emails.txt TTS.pages tts-telegram.txt TTS.warc.gz mask, target curl-run-all.sh discourse.mozilla.org html-to-markdown.sh ordered-posts ordered-posts~ TTS.cdx tts.commands tts-emails.txt TTS.pages tts-telegram.txt TTS.warc.gz mask, reduction='none')<br/> loss = loss.mul(out_weights.to(loss.device)).sum()<br/> else:<br/> mask = mask.expand_as(x)<br/> loss = functional.mse_loss(<br/> x curl-run-all.sh discourse.mozilla.org html-to-markdown.sh ordered-posts ordered-posts~ TTS.cdx tts.commands tts-emails.txt TTS.pages tts-telegram.txt TTS.warc.gz mask, target curl-run-all.sh discourse.mozilla.org html-to-markdown.sh ordered-posts ordered-posts~ TTS.cdx tts.commands tts-emails.txt TTS.pages tts-telegram.txt TTS.warc.gz mask, reduction='sum')<br/> loss = loss / mask.sum()<br/> return loss<br/> <br/> class AttentionEntropyLoss(nn.Module):<br/> # pylint: disable=R0201<br/> def forward(self, align):<br/> '''<br/> Forces attention to be more decisive by penalizing<br/> soft attention weights<br/> <br/> TODO: arguments<br/> TODO: unit_test<br/> '''<br/> entropy = torch.distributions.Categorical(probs=align).entropy()<br/> ```<br/> <br/> What is the expected tensor size?<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/how-to-use-attentionentropyloss]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [April 15, 2020, 9:41am]<br/> <br/> I've used it before but it is not active in the current code base. If<br/> you like you need to run it over the attention alignment vector.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> Had<br/> <DATE> [April 15, 2020, 5:49pm]<br/> <br/> I want to make alignments for fastspeech train.<br/> <br/> So every text token should consist with mel spectrogram frame. *slash* <br/> My alignment looks like this with multi head attention. *slash* <br/> It tend to skip whitespaces and other punctuation. *slash* <br/> Default tacotron is not perfect too. *slash* <br/> <br/> So I want to try this function to make the strict token-to-token<br/> alignment.<br/> <br/> Is it fine for *slash* [batch, width, height *slash* ]? *slash* <br/> Or I need to run it for every sample like *slash* [loss(i) for I in batch *slash* ]? *slash* <br/> Maybe I need something else?<br/> <br/> Thanks for attention.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [April 16, 2020, 10:06am]<br/> <br/> entropy loss would not do that. It works more like a guided attention<br/> loss which forces the alignment to be more diagonal.<br/> <br/> Alignment looks quite noisy. Pls share the config.json so I can guess<br/> better.<br/> <br/> [Archived Post]<br/>
